### DISCUSSION TITLE
Does VS Code IBM i have an outline like RDI does?

### DISCUSSION THREAD

**Discussion by WildWabbit13** [codefori/vscode-ibmi#1178] [IBM i] [answered]
The developers where I work all use RDI but I'm interested in this editor and trying it out.  One of our favorite RDI functions is the Outline window.

When I first brought it up there was some very small indistinguishable code on the far right.  I closed that.  I have the secondary sidebar up instead with outline dragged and dropped into it.  I get the error message: “The active editor cannot provide outline information.”  The editor is IBM i.  Does VS Code IBM i have the outline function that RDI has or something similar to it?  If so, how do I bring it up?  In viewing the videos on YouTube for IBM i it was not shown.

**Comment by worksofliam** [comment, marked as answer]
It does.

Ensure you have the RPGLE language tools installed: https://marketplace.visualstudio.com/items?itemName=HalcyonTechLtd.vscode-rpgle

**Reply by WildWabbit13** [reply]
I didn't.  It works fine now so thank you so much for replying.  One other question.  Is there a way of turning off insert when typing?  Thanks.

**Reply by chrjorgensen** [reply]
@WildWabbit13 You can [search the market place](https://marketplace.visualstudio.com/search?term=overtype&target=VSCode&category=All%20categories&sortBy=Relevance) for extensions, that will provide you with the desired functionality - I use [this one](https://marketplace.visualstudio.com/items?itemName=DrMerfy.overtype), since it's actively maintained.

**Comment by JohnJ49** [comment]
I'm having the same issue, loading the RPGLE language tools extension has not helped.

**Reply by sebjulliand** [reply]
Is the Outline view just empty or do you have any error in the Code for IBM i output and/or the console?

**Reply by JohnJ49** [reply]
Yes, should have added that. The error code is "No symbols found in document sysdoc.sqlrpgle"

**Reply by JohnJ49** [reply]
That's in the outline panel.

**Reply by sebjulliand** [reply]
Right, so maybe you can share the Code for IBM i output as well:
![image](https://user-images.githubusercontent.com/11096890/229278273-632ab984-377c-413a-ac57-b039d47b388f.png)

Open your RPG file and then copy and paste the content of that output. Maybe there is something useful in there.

You also have a look at the console after the file is opened, to see if there are any error there:
1. Open the `Help` menu and choose `Toggle Developer Tools`
![image](https://user-images.githubusercontent.com/11096890/229278462-510715f4-503d-4f8d-b3b1-cffe73dd7e3e.png)
2. Then select the `Console` tab
![image](https://user-images.githubusercontent.com/11096890/229278498-c17a3875-9b0a-4864-9160-425228402c43.png)
3. In the console, use `Ctrl+a` then `Ctrl+c` to copy the whole console content and paste it here as well.

Thanks!

**Comment by JohnJ49** [comment]
Here is a small test program that shows the issue. By the way I am using
PUB400.

[image: image.png]

Console output.
 INFO Electron sandbox mode is enabled!
log.ts:431  WARN Via 'product.json#extensionEnabledApiProposals' extension
'ms-vscode.vscode-selfhost-test-provider' wants API proposal
'testContinuousRun' but that proposal DOES NOT EXIST. Likely, the proposal
has been finalized (check 'vscode.d.ts') or was abandoned.
log.ts:431  WARN Via 'product.json#extensionEnabledApiProposals' extension
'ms-vscode.remote-repositories' wants API proposal
'contribEditorGutterMenu' but that proposal DOES NOT EXIST. Likely, the
proposal has been finalized (check 'vscode.d.ts') or was abandoned.
log.ts:431  WARN Via 'product.json#extensionEnabledApiProposals' extension
'github.vscode-pull-request-github' wants API proposal
'contribEditorGutterMenu' but that proposal DOES NOT EXIST. Likely, the
proposal has been finalized (check 'vscode.d.ts') or was abandoned.
log.ts:431  WARN Via 'product.json#extensionEnabledApiProposals' extension
'ms-toolsai.jupyter' wants API proposal 'notebookControllerKind' but that
proposal DOES NOT EXIST. Likely, the proposal has been finalized (check
'vscode.d.ts') or was abandoned.
log.ts:421  INFO [bitlang.cobol]: Command
`workbench.action.closeEditorsToTheLeft` appears multiple times in the
`commands` section.
log.ts:441   ERR [Extension Host] (node:15420) [DEP0005]
DeprecationWarning: Buffer() is deprecated due to security and usability
issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or
Buffer.from() methods instead.
(Use `Code --trace-deprecation ...` to show where the warning was created)
console.ts:137 [Extension Host] (node:15420) [DEP0005] DeprecationWarning:
Buffer() is deprecated due to security and usability issues. Please use the
Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead.
(Use `Code --trace-deprecation ...` to show where the warning was created)
T @ console.ts:137
console.ts:137 [Extension Host] Congratulations, your extension
"vscode-ibmi-walkthroughs" is now active!
log.ts:421  INFO [perf] Render performance baseline is 142ms
console.ts:137 [Extension Host] Congratulations, your extension
"code-for-ibmi" is now active!
console.ts:137 [Extension Host] Congratulations, your extension
"vscode-db2i" is now active!
console.ts:137 [Extension Host] Congratulations, your extension
"vscode-ibmi-notebooks" is now active!
console.ts:137 [Extension Host] CPD0032: Not authorized to command CRTLIB
in library *LIBL.
CPF0006: Errors occurred in command.
console.ts:137 [Extension Host] /QSYS.lib/JJANDREAU1.lib/GENCMDXML.PGM not
found
TMScopeRegistry.ts:45 Overwriting grammar scope name to file mapping for
scope source.sql.
Old grammar file:
file:///c%3A/Users/John/AppData/Local/Programs/Microsoft%20VS%20Code/resources/app/extensions/sql/syntaxes/sql.tmLanguage.json.
New grammar file:
file:///c%3A/Users/John/.vscode/extensions/mtxr.sqltools-0.27.1/language/sql.tmLanguage.json
register @ TMScopeRegistry.ts:45
TMScopeRegistry.ts:45 Overwriting grammar scope name to file mapping for
scope source.rpg.
Old grammar file:
file:///c%3A/Users/John/.vscode/extensions/barrettotte.ibmi-languages-0.6.10/syntaxes/rpg.tmLanguage.json.
New grammar file:
file:///c%3A/Users/John/.vscode/extensions/nielsliisberg.rpg-0.2.1/syntaxes/rpg.tmLanguage.json
register @ TMScopeRegistry.ts:45
TMScopeRegistry.ts:45 Overwriting grammar scope name to file mapping for
scope source.rpgle.
Old grammar file:
file:///c%3A/Users/John/.vscode/extensions/barrettotte.ibmi-languages-0.6.10/syntaxes/rpgle.tmLanguage.json.
New grammar file:
file:///c%3A/Users/John/.vscode/extensions/spinenetsystems.rpgle-1.1.16/syntaxes/rpg.tmLanguage.json
register @ TMScopeRegistry.ts:45
languageConfigurationExtensionPoint.ts:165 [rpg]: language configuration:
expected `comments.lineComment` to be a string.
k @ languageConfigurationExtensionPoint.ts:165
languageConfigurationExtensionPoint.ts:173 [rpg]: language configuration:
expected `comments.blockComment` to be an array of two strings.
k @ languageConfigurationExtensionPoint.ts:173

**Reply by sebjulliand** [reply]
Seems the image you tried to attach didn't make it through.

That's odd though...I just gave it a try, from a GitHub Codespace, with the RPGLE extension loaded and connected to PUB400; and even an RPGLE file opened from the file explorer shows the outline.

Can you check if the extension is both loaded and enabled?
![image](https://user-images.githubusercontent.com/11096890/229289801-2abaf02c-e542-451e-95f6-09d4aa85751d.png)

**Comment by JohnJ49** [comment]
It is loaded and enabled.

**Reply by sebjulliand** [reply]
I'd be curious to see the RPG file that causes the outline to remain empty. Thanks!

**Comment by JohnJ49** [comment]
Any RPG file that I open does not show the outline view.
In MiWorkplace the outline view shows up. Hopefully this screen print shows
up.

[image: image.png]

**Reply by sebjulliand** [reply]
Nope, it seems replying to the e-mail doesn't allow for images to be attached.

**Comment by JohnJ49** [comment]
I guess I'll remove vscode and do a fresh install. I should be able to get
to that on Monday.

**Comment by JohnJ49** [comment]
Hi
Finally got it working. I had to remove all extensions, remove VS Code, and
then reload everything.

Thanks for your help.

John

**Reply by sebjulliand** [reply]
Anytime! I'm glad it finally works.

**Comment by JohnJ49** [comment]
I have found an extension that stops the outline view from working. That is
RRPG by SpineNetSystems.
I do have a question about the outline view. Is it supposed to show where
items listed in the view are used? Right now all I get is the definition,
not where used.

Thanks

John

**Reply by worksofliam** [reply]
Sadly I don't know much about that extension, though there should be no reason they (Code for IBM i and RRPG) can't work together. I imagine RPGLE (from us) and RRPG can't work together since they probably try to provide the same thing.

**Comment by JohnJ49** [comment]
The description doesn't say anything about providing an outline view.
And it seems to me that the outline view should show "where used" for each
item in the view. Maybe it does and my install still has an issue.

**Comment by Gregory1701** [comment]
My outline view worked and now it shows 'No symbols found in document...' on any document I open.  Here is a sample RPG:
**Free
Ctl-Opt Option(*Nodebugio:*Srcstmt);
Ctl-Opt ActGrp(*New);

PerformFibonacciSequence();

*InLr = *On;
Return;

Dcl-Proc PerformFibonacciSequence;

Dcl-pi *n;
End-pi;

Dcl-s var_1 Uns(3) Inz(0);
Dcl-s var_2 Uns(3) Inz(1);
Dcl-s var_3 Uns(3) Inz;
Dcl-s count Uns(3) Inz;

  Dow count <= 11;
    Dsply var_1;
    var_3 = var_1 + var_2;
    var_1 = var_2;
    var_2 = var_3;
    count += 1;
  Enddo;

End-Proc;

Here is my console information:
workbench.desktop.main.js:sourcemap:624  WARN Via 'product.json#extensionEnabledApiProposals' extension 'ms-vscode.vscode-selfhost-test-provider' wants API proposal 'testContinuousRun' but that proposal DOES NOT EXIST. Likely, the proposal has been finalized (check 'vscode.d.ts') or was abandoned.
workbench.desktop.main.js:sourcemap:624  INFO Invoking resolveAuthority(wsl)...
workbench.desktop.main.js:sourcemap:624  INFO [LocalProcess0][resolveAuthority(wsl,1)][0ms] obtaining proxy...
workbench.desktop.main.js:sourcemap:624  INFO [LocalProcess0][resolveAuthority(wsl,1)][1ms] invoking...
workbench.desktop.main.js:sourcemap:624  INFO [LocalProcess0][resolveAuthority(wsl,1)][1003ms] waiting...
workbench.desktop.main.js:sourcemap:624  INFO [LocalProcess0][resolveAuthority(wsl,1)][2002ms] waiting...
workbench.desktop.main.js:sourcemap:624  INFO [LocalProcess0][resolveAuthority(wsl,1)][3006ms] waiting...
workbench.desktop.main.js:sourcemap:624  INFO [LocalProcess0][resolveAuthority(wsl,1)][3370ms] returned 127.0.0.1:57866
workbench.desktop.main.js:sourcemap:624  INFO resolveAuthority(wsl) returned '127.0.0.1:57866' after 3370 ms
workbench.desktop.main.js:sourcemap:624  INFO Creating a socket (renderer-Management-3398c46c-7b9f-48da-b04c-769298a330f9)...
workbench.desktop.main.js:sourcemap:624  INFO Creating a socket (renderer-ExtensionHost-761635ac-cfa0-4a73-8344-80527cfd2d96)...
workbench.desktop.main.js:sourcemap:624  INFO Creating a socket (renderer-Management-3398c46c-7b9f-48da-b04c-769298a330f9) was successful after 546 ms.
workbench.desktop.main.js:sourcemap:624  INFO Creating a socket (renderer-ExtensionHost-761635ac-cfa0-4a73-8344-80527cfd2d96) was successful after 1259 ms.
workbench.desktop.main.js:sourcemap:624  INFO [perf] Render performance baseline is 67ms
workbench.desktop.main.js:sourcemap:1842 Overwriting grammar scope name to file mapping for scope source.rpg.
Old grammar file: file:///c%3A/Users/PGMgregory/.vscode/extensions/barrettotte.ibmi-languages-0.6.12/syntaxes/rpg.tmLanguage.json.
New grammar file: file:///c%3A/Users/PGMgregory/.vscode/extensions/nielsliisberg.rpg-0.2.1/syntaxes/rpg.tmLanguage.json
register @ workbench.desktop.main.js:sourcemap:1842
workbench.desktop.main.js:sourcemap:1842 Overwriting grammar scope name to file mapping for scope source.yaml.
Old grammar file: file:///c%3A/Users/PGMgregory/AppData/Local/Programs/Microsoft%20VS%20Code/resources/app/extensions/yaml/syntaxes/yaml.tmLanguage.json.
New grammar file: vscode-remote://wsl%2Bfedora/home/pgmgregory/.vscode-server/extensions/redhat.vscode-yaml-1.12.2/syntaxes/yaml.tmLanguage.json
register @ workbench.desktop.main.js:sourcemap:1842
workbench.desktop.main.js:sourcemap:97 [Extension Host] Congratulations, your extension "vscode-ibmi-walkthroughs" is now active!
workbench.desktop.main.js:sourcemap:97 [Extension Host] Congratulations, your extension "code-for-ibmi" is now active!
workbench.desktop.main.js:sourcemap:97 [Extension Host] Developer environment: false
workbench.desktop.main.js:sourcemap:97 [Extension Host] Congratulations, your extension "vscode-db2i" is now active!
workbench.desktop.main.js:sourcemap:97 [Extension Host] Congratulations, your extension "vscode-ibmi-notebooks" is now active!
workbench.desktop.main.js:sourcemap:624   ERR [Extension Host] (node:20772) [DEP0005] DeprecationWarning: Buffer() is deprecated due to security and usability issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead.
(Use `node --trace-deprecation ...` to show where the warning was created)
workbench.desktop.main.js:sourcemap:97 [Extension Host] (node:20772) [DEP0005] DeprecationWarning: Buffer() is deprecated due to security and usability issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead.
(Use `node --trace-deprecation ...` to show where the warning was created)
I @ workbench.desktop.main.js:sourcemap:97
workbench.desktop.main.js:sourcemap:97 [Extension Host] CPF2111: Library ILEDITOR already exists.
workbench.desktop.main.js:sourcemap:97 [Extension Host] Using new temp: /tmp/vscodetemp-O_4bUNrXDS

**Reply by worksofliam** [reply]
There are no errors in your log and it works ok for me.

<img width="808" alt="image" src="https://github.com/halcyon-tech/vscode-ibmi/assets/3708366/2dd632fb-c98a-43c0-84bd-43a996fbb26f">

Ensure the extension is correct: `.rpgle` or `.sqlrpgle`.

**Reply by Gregory1701** [reply]
Hmm... it is a .rpgle type source member.  The only difference I'm seeing between yours and mine is I'm using dark theme?
![image](https://github.com/halcyon-tech/vscode-ibmi/assets/132368707/f03f8490-7ded-407e-a958-0cfa39c0806f)

**Reply by Gregory1701** [reply]
Yeah, aside from hurting my eyes, switching to light theme didn't do anything for me. 😎

**Reply by worksofliam** [reply]
I bet the issue is due to the fact you have `nielsliisberg.rpg` installed. I believe there is a clash between two extensions registering support for rpgle.

**Reply by Gregory1701** [reply]
Ah, hah!  That was it.  I uninstalled that extension and now my Outline view is working.  Thanks!
